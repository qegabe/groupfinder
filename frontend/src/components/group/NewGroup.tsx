import React, { useState } from "react";
import { Box, Typography } from "@mui/material";
import GroupFinderApi from "../../api";
import GroupForm from "./GroupForm";

const INITIAL_STATE: GroupFormData = {
  title: "",
  description: "",
  startTime: null,
  endTime: null,
  address: "",
  cityData: null,
  isPrivate: false,
  maxMembers: 10,
};

function NewGroup() {
  const [formData, setFormData] = useState<GroupFormData>(INITIAL_STATE);

  async function submit() {
    await GroupFinderApi.createGroup(formData);
  }

  return (
    <Box sx={{ display: "grid", justifyItems: "center" }}>
      <Typography variant="h3">Create New Group</Typography>
      <GroupForm
        initialState={INITIAL_STATE}
        formData={formData}
        setFormData={setFormData}
        submit={submit}
        returnPath="/groups"
        shouldReturn={true}
        buttons={{ submit: "Create", cancel: "Cancel" }}
      />
    </Box>
  );
}

export default NewGroup;
