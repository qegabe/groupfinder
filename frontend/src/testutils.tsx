import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { RootState, setupStore } from "./store";
import { Provider } from "react-redux";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { LocalizationProvider } from "@mui/x-date-pickers";
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import { PreloadedState } from "@reduxjs/toolkit";

interface CustomRenderOptions {
  currentRoute?: string;
  routePath?: string;
  preloadedState?: PreloadedState<RootState>;
}

function customRender(
  ui: ReactElement,
  cOptions: CustomRenderOptions = {},
  options?: Omit<RenderOptions, "wrapper">
) {
  const store = setupStore(cOptions.preloadedState);
  function AllTheProviders({ children }: { children: React.ReactNode }) {
    return (
      <Provider store={store}>
        <MemoryRouter
          initialEntries={
            cOptions.currentRoute ? [cOptions.currentRoute] : undefined
          }>
          <LocalizationProvider dateAdapter={AdapterDayjs}>
            {cOptions.routePath ? (
              <Routes>
                <Route path={cOptions.routePath} element={children} />
              </Routes>
            ) : (
              children
            )}
          </LocalizationProvider>
        </MemoryRouter>
      </Provider>
    );
  }
  return render(ui, { wrapper: AllTheProviders, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
