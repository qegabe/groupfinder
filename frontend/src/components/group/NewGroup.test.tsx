import React from "react";
import { render, waitFor, screen, fireEvent, within } from "../../testutils";
import "@testing-library/jest-dom";
import GroupFinderApi from "../../api";
import dayjs from "dayjs";
import NewGroup from "./NewGroup";

const mockCreateGroup = jest
  .spyOn(GroupFinderApi, "createGroup")
  .mockImplementation(async () => {});

it("renders", async () => {
  render(<NewGroup />);
  expect(screen.getByText("Create New Group")).toBeInTheDocument();
});

it("matches snapshot", async () => {
  const { asFragment } = render(<NewGroup />);
  expect(asFragment()).toMatchSnapshot();
});

it("allows user to create group", async () => {
  render(<NewGroup />);
  const titleInput = screen.getByLabelText("Title *");
  const startTimeInput = screen.getByLabelText("Start Time");
  const endTimeInput = screen.getByLabelText("End Time");
  const createButton = screen.getByText("Create");
  fireEvent.change(titleInput, { target: { value: "Test" } });
  fireEvent.change(startTimeInput, {
    target: { value: "06/23/2023 12:00 PM" },
  });
  fireEvent.change(endTimeInput, {
    target: { value: "06/23/2023 01:00 PM" },
  });
  fireEvent.click(createButton);

  waitFor(() => {
    expect(screen.getByText("Create New Group")).not.toBeInTheDocument();
  });
  expect(mockCreateGroup).toBeCalledWith({
    title: "Test",
    description: "",
    startTime: dayjs("2023-06-23T16:00:00.000Z"),
    endTime: dayjs("2023-06-23T17:00:00.000Z"),
    address: "",
    cityData: null,
    isPrivate: false,
    maxMembers: 10,
  });
});
