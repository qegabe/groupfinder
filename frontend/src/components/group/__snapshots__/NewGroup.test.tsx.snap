// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-5mr1lz"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-14jhf3t-MuiTypography-root"
    >
      Create New Group
    </p>
    <form
      autocomplete="off"
      class="MuiBox-root css-bxifig"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1nin81k-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="title"
          id="title-label"
        >
          Title
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="title"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Title *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1nin81k-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="description"
          id="description-label"
        >
          Description
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            id="description"
            style="height: 0px; overflow: hidden;"
          />
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
            tabindex="-1"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Description
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-5h9f05-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r9:"
          id=":r9:-label"
        >
          Start Time
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r9:"
            inputmode="text"
            placeholder="MM/DD/YYYY hh:mm aa"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Start Time
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-5h9f05-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":rb:"
          id=":rb:-label"
        >
          End Time
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            id=":rb:"
            inputmode="text"
            placeholder="MM/DD/YYYY hh:mm aa"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                End Time
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiBox-root css-1impqaw"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
          >
            <span
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-checked css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                id="isOnline"
                type="checkbox"
              />
              <span
                class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            Online
          </span>
        </label>
      </div>
      <div
        class="MuiBox-root css-1ebhyd8"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
          >
            <span
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
            >
              <input
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                id="isPrivate"
                type="checkbox"
              />
              <span
                class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            Private Group?
          </span>
        </label>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="maxMembers"
            id="maxMembers-label"
          >
            Maximum Members
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="maxMembers"
              min="0"
              type="number"
              value="10"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Maximum Members
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-a1o1ks"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-bh137t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Create
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <a
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1l9sosn-MuiButtonBase-root-MuiButton-root"
          href="/groups"
          tabindex="0"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
      </div>
      <div
        class="MuiBox-root css-144iql5"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1m6glrx-MuiTypography-root"
        >
          City information provided by: 
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1bhi0za-MuiTypography-root-MuiLink-root"
            href="https://simplemaps.com/data/world-cities"
          >
            simplemaps
          </a>
        </p>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
